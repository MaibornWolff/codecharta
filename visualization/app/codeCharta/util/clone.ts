import { klona } from "klona/json"

export function clone<T>(content: T): T {
	return klona(content)
}
