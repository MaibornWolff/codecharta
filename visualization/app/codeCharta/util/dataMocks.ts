import { CCFile, CodeMapNode, Edge, MetricData, Node, Settings } from "../codeCharta.model"
import { CodeMapBuilding } from "../ui/codeMap/rendering/codeMapBuilding"
import * as THREE from "three"

export const VALID_NODE: CodeMapNode = {
	name: "root",
	attributes: {},
	type: "Folder",
	children: [
		{
			name: "big leaf",
			type: "File",
			attributes: { RLOC: 100, Functions: 10, MCC: 1 },
			link: "http://www.google.de"
		},
		{
			name: "Parent Leaf",
			type: "Folder",
			attributes: {},
			children: [
				{
					name: "small leaf",
					type: "File",
					attributes: { RLOC: 30, Functions: 100, MCC: 100 }
				},
				{
					name: "other small leaf",
					type: "File",
					attributes: { RLOC: 70, Functions: 1000, MCC: 10 }
				}
			]
		}
	]
}

export const VALID_NODE_WITH_PATH: CodeMapNode = {
	name: "root",
	attributes: {},
	type: "Folder",
	path: "/root",
	children: [
		{
			name: "big leaf",
			type: "File",
			path: "/root/big leaf",
			attributes: { RLOC: 100, Functions: 10, MCC: 1 },
			link: "http://www.google.de"
		},
		{
			name: "Parent Leaf",
			type: "Folder",
			attributes: {},
			path: "/root/Parent Leaf",
			children: [
				{
					name: "small leaf",
					type: "File",
					path: "/root/Parent Leaf/small leaf",
					attributes: { RLOC: 30, Functions: 100, MCC: 100 }
				},
				{
					name: "other small leaf",
					type: "File",
					path: "/root/Parent Leaf/other small leaf",
					attributes: { RLOC: 70, Functions: 1000, MCC: 10 }
				},
				{
					name: "empty folder",
					type: "Folder",
					path: "/root/Parent Leaf/empty folder",
					attributes: {},
					children: []
				}
			]
		}
	]
}

export const VALID_EDGES: Edge[] = [
	{
		fromNodeName: "/root/big leaf",
		toNodeName: "/root/Parent Leaf/small leaf",
		attributes: {
			pairingRate: 89,
			avgCommits: 34
		}
	},
	{
		fromNodeName: "/root/sample1 only leaf",
		toNodeName: "/root/Parent Leaf/small leaf",
		attributes: {
			pairingRate: 89,
			avgCommits: 34
		}
	}
]

export const VALID_EDGE: Edge = {
	fromNodeName: "/root/big leaf",
	toNodeName: "/root/Parent Leaf/small leaf",
	attributes: {
		pairingRate: 89,
		avgCommits: 34
	}
}

export const TEST_FILE_CONTENT = {
	fileName: "noFileName",
	projectName: "Sample Map",
	apiVersion: "1.1",
	nodes: [VALID_NODE]
}

export const TEST_FILE_DATA: CCFile = {
	fileMeta: {
		fileName: "file",
		projectName: "Sample Map",
		apiVersion: "1.1"
	},
	map: VALID_NODE,
	settings: {
		fileSettings: {
			attributeTypes: {},
			blacklist: [],
			edges: VALID_EDGES,
			markedPackages: []
		}
	}
}

export const TEST_FILE_WITH_PATHS: CCFile = {
	fileMeta: {
		fileName: "fileA",
		projectName: "Sample Project",
		apiVersion: "1.1"
	},
	map: {
		name: "root",
		type: "Folder",
		path: "/root",
		attributes: {},
		children: [
			{
				name: "big leaf",
				type: "File",
				path: "/root/big leaf",
				attributes: { rloc: 100, functions: 10, mcc: 1 },
				link: "http://www.google.de"
			},
			{
				name: "Parent Leaf",
				type: "Folder",
				attributes: {},
				path: "/root/Parent Leaf",
				children: [
					{
						name: "small leaf",
						type: "File",
						path: "/root/Parent Leaf/small leaf",
						attributes: { rloc: 30, functions: 100, mcc: 100 }
					},
					{
						name: "other small leaf",
						type: "File",
						path: "/root/Parent Leaf/other small leaf",
						attributes: { rloc: 70, functions: 1000, mcc: 10 }
					},
					{
						name: "empty folder",
						type: "Folder",
						path: "/root/Parent Leaf/empty folder",
						attributes: {},
						children: []
					}
				]
			}
		]
	},
	settings: {
		fileSettings: {
			attributeTypes: {},
			blacklist: [],
			edges: VALID_EDGES,
			markedPackages: []
		}
	}
}

export const TEST_DELTA_MAP_A: CCFile = {
	fileMeta: {
		fileName: "fileA",
		projectName: "Sample Project",
		apiVersion: "1.1"
	},
	map: {
		name: "root",
		type: "Folder",
		attributes: {},
		children: [
			{
				name: "big leaf",
				type: "File",
				attributes: { rloc: 100, functions: 10, mcc: 1 },
				link: "http://www.google.de"
			},
			{
				name: "Parent Leaf",
				type: "Folder",
				attributes: {},
				children: [
					{
						name: "small leaf",
						type: "File",
						attributes: { rloc: 30, functions: 100, mcc: 100 }
					},
					{
						name: "other small leaf",
						type: "File",
						attributes: { rloc: 70, functions: 1000, mcc: 10 }
					}
				]
			}
		]
	},
	settings: {
		fileSettings: {
			attributeTypes: {},
			blacklist: [],
			edges: VALID_EDGES,
			markedPackages: []
		}
	}
}

export const TEST_DELTA_MAP_B: CCFile = {
	fileMeta: {
		fileName: "fileB",
		projectName: "Sample Project",
		apiVersion: "1.1"
	},
	map: {
		name: "root",
		type: "Folder",
		attributes: {},
		children: [
			{
				name: "big leaf",
				type: "File",
				attributes: { rloc: 20, functions: 10, mcc: 1 },
				link: "http://www.google.de"
			},
			{
				name: "additional leaf",
				type: "File",
				attributes: { rloc: 10, functions: 11, mcc: 5 },
				link: "http://www.google.de"
			},
			{
				name: "Parent Leaf",
				type: "Folder",
				attributes: {},
				children: [
					{
						name: "small leaf",
						type: "File",
						attributes: { rloc: 30, functions: 100, mcc: 100, more: 20 }
					},
					{
						name: "other small leaf",
						type: "File",
						attributes: { rloc: 70, functions: 1000 }
					},
					{
						name: "big leaf",
						type: "File",
						attributes: { rloc: 20, functions: 10, mcc: 1 },
						link: "http://www.google.de"
					}
				]
			}
		]
	},
	settings: {
		fileSettings: {
			attributeTypes: {},
			blacklist: [],
			edges: VALID_EDGES,
			markedPackages: []
		}
	}
}

export const TEST_FILE_DATA_DOWNLOADED = {
	apiVersion: "1.1",
	attributeTypes: {},
	blacklist: [],
	edges: [
		{
			attributes: { avgCommits: 34, pairingRate: 89 },
			fromNodeName: "/root/big leaf",
			toNodeName: "/root/Parent Leaf/small leaf"
		},
		{
			attributes: { avgCommits: 34, pairingRate: 89 },
			fromNodeName: "/root/sample1 only leaf",
			toNodeName: "/root/Parent Leaf/small leaf"
		}
	],
	fileName: "file_2018-12-14_9-39.cc.json",
	nodes: [
		{
			attributes: {},
			children: [
				{
					attributes: { Functions: 10, MCC: 1, RLOC: 100 },
					link: "http://www.google.de",
					name: "big leaf",
					type: "File"
				},
				{
					attributes: {},
					children: [
						{
							attributes: { Functions: 100, MCC: 100, RLOC: 30 },
							name: "small leaf",
							type: "File"
						},
						{
							attributes: { Functions: 1000, MCC: 10, RLOC: 70 },
							name: "other small leaf",
							type: "File"
						}
					],
					name: "Parent Leaf",
					type: "Folder"
				}
			],
			name: "root",
			type: "Folder"
		}
	],
	projectName: "Sample Map"
}

export const SETTINGS: Settings = {
	fileSettings: { attributeTypes: {}, blacklist: [], edges: [], markedPackages: [] },
	dynamicSettings: {
		areaMetric: "rloc",
		heightMetric: "mcc",
		colorMetric: "mcc",
		focusedNodePath: "/root",
		searchedNodePaths: [],
		searchPattern: "",
		margin: 48,
		neutralColorRange: { flipped: false, from: 19, to: 67 }
	},
	appSettings: {
		amountOfTopLabels: 31,
		scaling: new THREE.Vector3(1, 1.8, 1),
		camera: new THREE.Vector3(0, 300, 1000),
		deltaColorFlipped: false,
		enableEdgeArrows: true,
		hideFlatBuildings: true,
		maximizeDetailPanel: false,
		invertHeight: true,
		dynamicMargin: true,
		isWhiteBackground: false,
		whiteColorBuildings: true,
		mapColors: {
			positive: "#69AE40",
			neutral: "#ddcc00",
			negative: "#820E0E",
			selected: "#EB8319",
			defaultC: "#89ACB4",
			positiveDelta: "#69FF40",
			negativeDelta: "#ff0E0E",
			base: "#666666",
			flat: "#AAAAAA",
			lightGrey: "#DDDDDD",
			angularGreen: "#00BFA5",
			markingColors: ["#FF1D8E", "#1d8eff", "#1DFFFF", "#8eff1d", "#8e1dff", "#FFFF1D"]
		}
	},
	treeMapSettings: { mapSize: 500 }
}

export const DEFAULT_SETTINGS: Settings = {
	appSettings: {
		amountOfTopLabels: 1,
		camera: new THREE.Vector3(0, 300, 1000),
		deltaColorFlipped: false,
		dynamicMargin: true,
		enableEdgeArrows: true,
		hideFlatBuildings: true,
		invertHeight: false,
		isWhiteBackground: false,
		mapColors: {
			angularGreen: "#00BFA5",
			base: "#666666",
			defaultC: "#89ACB4",
			flat: "#AAAAAA",
			lightGrey: "#DDDDDD",
			markingColors: ["#FF1D8E", "#1d8eff", "#1DFFFF", "#8eff1d", "#8e1dff", "#FFFF1D"],
			negative: "#820E0E",
			negativeDelta: "#ff0E0E",
			neutral: "#ddcc00",
			positive: "#69AE40",
			positiveDelta: "#69FF40",
			selected: "#EB8319"
		},
		maximizeDetailPanel: false,
		scaling: new THREE.Vector3(1, 1, 1),
		whiteColorBuildings: false
	},
	dynamicSettings: {
		areaMetric: null,
		colorMetric: null,
		focusedNodePath: "",
		heightMetric: null,
		margin: null,
		neutralColorRange: {
			flipped: false,
			from: null,
			to: null
		},
		searchPattern: "",
		searchedNodePaths: []
	},
	fileSettings: { attributeTypes: {}, blacklist: [], edges: [], markedPackages: [] },
	treeMapSettings: { mapSize: 500 }
}

export const TEST_NODE_ROOT: Node = {
	name: "root",
	width: 1,
	height: 2,
	length: 3,
	depth: 4,
	x0: 5,
	z0: 6,
	y0: 7,
	isLeaf: true,
	deltas: { a: 1, b: 2 },
	attributes: { a: 20, b: 15 },
	children: [],
	parent: undefined,
	heightDelta: 10,
	visible: true,
	path: "root",
	origin: "root",
	link: "NO_LINK",
	markingColor: "0x000000",
	flat: false
}

export const TEST_NODE_LEAF: Node = {
	name: "root/big leaf",
	width: 1,
	height: 2,
	length: 3,
	depth: 4,
	x0: 5,
	z0: 6,
	y0: 7,
	isLeaf: true,
	deltas: { a: 1, b: 2 },
	attributes: { a: 20, b: 15 },
	children: [],
	parent: undefined,
	heightDelta: 20,
	visible: true,
	path: "root/big leaf",
	origin: "root",
	link: "NO_LINK",
	markingColor: "0xFFFFFF",
	flat: false
}

export const CODE_MAP_BUILDING: CodeMapBuilding = new CodeMapBuilding(
	1,
	new THREE.Box3(),
	TEST_NODE_ROOT,
	DEFAULT_SETTINGS.appSettings.mapColors.neutral
)

export const METRIC_DATA: MetricData[] = [
	{ name: "mcc", maxValue: 1, availableInVisibleMaps: true },
	{ name: "rloc", maxValue: 2, availableInVisibleMaps: true }
]
