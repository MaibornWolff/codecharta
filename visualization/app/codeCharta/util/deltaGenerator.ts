import { CodeMapNode, CCFile, KeyValuePair, FileMeta } from "../codeCharta.model"
import { FileNameHelper } from "./fileNameHelper"
import { hierarchy } from "d3-hierarchy"
import packageJson from "../../../package.json"
import { CodeChartaService } from "../codeCharta.service"
import { getParent } from "./nodePathHelper"

export class DeltaGenerator {
	static createCodeMapFromHashMap(hashMapWithAllNodes: Map<string, CodeMapNode>) {
		let rootNode: CodeMapNode
		for (const [path, node] of hashMapWithAllNodes) {
			if (path === CodeChartaService.ROOT_PATH) {
				rootNode = node
			} else {
				const parentNode = getParent(hashMapWithAllNodes, path)
				parentNode.children.push(node)
			}
		}
		return rootNode
	}

	static getDeltaFile(referenceFile: CCFile, comparisonFile: CCFile) {
		const hashMapWithAllNodes = this.getHashMapWithAllNodes(referenceFile.map, comparisonFile.map)
		const map = this.createCodeMapFromHashMap(hashMapWithAllNodes)
		const fileMeta = this.getFileMetaData(referenceFile, comparisonFile)
		return this.getNewCCFileWithDeltas(map, fileMeta)
	}

	private static getHashMapWithAllNodes(referenceMap: CodeMapNode, comparisonMap: CodeMapNode) {
		const hashMapWithAllNodes: Map<string, CodeMapNode> = new Map()
		const referenceHashMap: Map<string, CodeMapNode> = new Map()

		// Get one side of the nodes
		for (const { data } of hierarchy(referenceMap)) {
			referenceHashMap.set(data.path, data)
		}

		// Combine both sides of the nodes
		for (const { data: comparisonNode } of hierarchy(comparisonMap)) {
			const referenceNode = referenceHashMap.get(comparisonNode.path)
			if (referenceNode) {
				if (referenceNode.children || comparisonNode.children) {
					referenceNode.children = []
				}
				referenceNode.deltas = this.getDeltaAttributeList(referenceNode.attributes, comparisonNode.attributes)
				// TODO: The attributes have to be consolidated to have a single set of
				// attributes instead of conflicting attributes. This applies to all
				// attributes and is not specific about the attributes from the
				// reference node.
				referenceNode.attributes = comparisonNode.attributes
			} else {
				if (comparisonNode.children) {
					comparisonNode.children = []
				}
				comparisonNode.deltas = { ...comparisonNode.attributes }
			}
			const node = referenceNode ?? comparisonNode
			hashMapWithAllNodes.set(node.path, node)
			referenceHashMap.delete(node.path)
		}

		// Add missing nodes
		for (const node of referenceHashMap.values()) {
			if (node.children) {
				node.children = []
			}
			node.deltas = {}
			for (const [key, value] of Object.entries(node.attributes)) {
				node.deltas[key] = -value
			}
			hashMapWithAllNodes.set(node.path, node)
		}

		return hashMapWithAllNodes
	}

	private static getDeltaAttributeList(referenceAttribute: KeyValuePair, comparisonAttribute: KeyValuePair) {
		const deltaAttribute: KeyValuePair = {}

		// TODO: All entries should have the combined attributes and deltas set,
		// even if they do not exist on one side. Calculate these attributes up
		// front. This operation is otherwise costly.
		for (const key of Object.keys(comparisonAttribute)) {
			deltaAttribute[key] = referenceAttribute[key] ? comparisonAttribute[key] - referenceAttribute[key] : comparisonAttribute[key]
		}

		for (const key of Object.keys(referenceAttribute)) {
			if (!comparisonAttribute[key]) {
				deltaAttribute[key] = -referenceAttribute[key]
			}
		}
		return deltaAttribute
	}

	private static getFileMetaData(referenceFile: CCFile, comparisonFile: CCFile): FileMeta {
		return {
			fileName: `delta_between_${FileNameHelper.withoutCCJsonExtension(
				referenceFile.fileMeta.fileName
			)}_and_${FileNameHelper.withoutCCJsonExtension(comparisonFile.fileMeta.fileName)}`,
			fileChecksum: `${referenceFile.fileMeta.fileChecksum};${comparisonFile.fileMeta.fileChecksum}`,
			apiVersion: packageJson.codecharta.apiVersion,
			projectName: `delta_between_${referenceFile.fileMeta.projectName}_and_${comparisonFile.fileMeta.projectName}`,
			exportedFileSize: referenceFile.fileMeta.exportedFileSize + comparisonFile.fileMeta.exportedFileSize
		}
	}

	private static getNewCCFileWithDeltas(rootNode: CodeMapNode, fileMeta: FileMeta): CCFile {
		return {
			map: rootNode,
			fileMeta,
			settings: {
				fileSettings: {
					edges: [],
					blacklist: [],
					attributeTypes: { nodes: {}, edges: {} },
					markedPackages: []
				}
			}
		}
	}
}
