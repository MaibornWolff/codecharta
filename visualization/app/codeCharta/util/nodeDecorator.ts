"use strict"
import * as d3 from "d3"
import { HierarchyNode } from "d3"
import {
	BlacklistItem,
	BlacklistType,
	CCFile,
	CodeMapNode,
	MetricData,
	FileMeta,
	EdgeMetricCount,
	KeyValuePair,
	AttributeTypes,
	AttributeTypeValue
} from "../codeCharta.model"
import { CodeMapHelper } from "./codeMapHelper"
import { MetricService } from "../state/metric.service"

export class NodeDecorator {
	public static decorateMap(map: CodeMapNode, fileMeta: FileMeta, metricData: MetricData[]) {
		this.decorateMapWithMissingObjects(map, fileMeta)
		this.decorateMapWithCompactMiddlePackages(map)
		this.decorateLeavesWithMissingMetrics(map, metricData)
	}

	public static preDecorateFile(file: CCFile) {
		this.decorateMapWithPathAttribute(file)
	}

	private static decorateMapWithCompactMiddlePackages(map: CodeMapNode) {
		const isEmptyMiddlePackage = current => {
			return (
				current &&
				current.children &&
				current.children.length === 1 &&
				current.children[0].children &&
				current.children[0].children.length > 0
			)
		}

		const rec = current => {
			if (isEmptyMiddlePackage(current)) {
				let child = current.children[0]
				current.children = child.children
				current.name += "/" + child.name
				current.path += "/" + child.name
				if (child.link) {
					current.link = child.link
				}
				current.attributes = child.attributes
				current.edgeAttributes = child.edgeAttributes
				current.deltas = child.deltas
				rec(current)
			} else if (current && current.children && current.children.length > 1) {
				for (let i = 0; i < current.children.length; i++) {
					rec(current.children[i])
				}
			}
		}

		if (map) {
			rec(map)
		}
	}

	private static decorateMapWithPathAttribute(file: CCFile) {
		if (file && file.map) {
			let root = d3.hierarchy<CodeMapNode>(file.map)
			root.each(node => {
				node.data.path =
					"/" +
					root
						.path(node)
						.map(x => x.data.name)
						.join("/")
			})
		}
		return file
	}

	private static decorateMapWithMissingObjects(map: CodeMapNode, fileMeta: FileMeta) {
		if (map) {
			let root = d3.hierarchy<CodeMapNode>(map)
			root.each(node => {
				node.data.visible = true
				node.data.attributes = !node.data.attributes ? {} : node.data.attributes
				node.data.edgeAttributes = !node.data.edgeAttributes ? {} : node.data.edgeAttributes
				Object.assign(node.data.attributes, { [MetricService.UNARY_METRIC]: 1 })
			})
		}
	}

	private static decorateLeavesWithMissingMetrics(map: CodeMapNode, metricData: MetricData[]) {
		if (map && metricData) {
			let root = d3.hierarchy<CodeMapNode>(map)
			root.leaves().forEach(node => {
				metricData.forEach(metric => {
					if (!node.data.attributes.hasOwnProperty(metric.name)) {
						node.data.attributes[metric.name] = 0
					}
				})
			})
		}
	}

	public static decorateParentNodesWithAggregatedAttributes(
		map: CodeMapNode,
		blacklist: BlacklistItem[],
		metricData: MetricData[],
		edgeMetricData: MetricData[],
		isDeltaState: boolean,
		attributeTypes: AttributeTypes
	) {
		if (map) {
			let root = d3.hierarchy<CodeMapNode>(map)
			root.each((node: HierarchyNode<CodeMapNode>) => {
				const leaves: HierarchyNode<CodeMapNode>[] = node
					.leaves()
					.filter(x => !CodeMapHelper.isBlacklisted(x.data, blacklist, BlacklistType.exclude))
				this.decorateNodeWithAggregatedChildrenMetrics(leaves, node, metricData, isDeltaState, attributeTypes)
				this.decorateNodeWithChildrenSumEdgeMetrics(leaves, node, edgeMetricData, attributeTypes)
			})
		}
		return map
	}

	private static decorateNodeWithAggregatedChildrenMetrics(
		leaves: HierarchyNode<CodeMapNode>[],
		node: HierarchyNode<CodeMapNode>,
		metricData: MetricData[],
		isDeltaState: boolean,
		attributeTypes: AttributeTypes
	) {
		metricData.forEach(metric => {
			if (node.data.children && node.data.children.length > 0) {
				node.data.attributes[metric.name] = this.aggregateLeafMetric(
					leaves.map(x => x.data.attributes),
					metric.name,
					attributeTypes
				)
				if (isDeltaState) {
					node.data.deltas[metric.name] = this.aggregateLeafMetric(leaves.map(x => x.data.deltas), metric.name, attributeTypes)
				}
			}
		})
	}

	private static decorateNodeWithChildrenSumEdgeMetrics(
		leaves: HierarchyNode<CodeMapNode>[],
		node: HierarchyNode<CodeMapNode>,
		edgeMetricData: MetricData[],
		attributeTypes: AttributeTypes
	) {
		edgeMetricData.forEach(edgeMetric => {
			if (node.data.children && node.data.children.length > 0) {
				node.data.edgeAttributes[edgeMetric.name] = this.aggregateLeafEdgeMetric(leaves, edgeMetric.name, attributeTypes)
			}
		})
	}

	private static aggregateLeafMetric(metrics: KeyValuePair[], metricName: string, attributeTypes: AttributeTypes): number {
		const metricValues: number[] = metrics.map(x => x[metricName]).filter(x => !!x)
		const attributeType = attributeTypes.nodes[metricName]

		if (metricValues.length == 0) {
			return 0
		}

		switch (attributeType) {
			case AttributeTypeValue.relative:
				return this.median(metricValues)
			case AttributeTypeValue.absolute:
			default:
				return metricValues.reduce((partialSum, a) => partialSum + a)
		}
	}

	private static aggregateLeafEdgeMetric(
		leaves: HierarchyNode<CodeMapNode>[],
		metricName: string,
		attributeTypes: AttributeTypes
	): EdgeMetricCount {
		const metricValues: EdgeMetricCount[] = leaves.map(x => x.data.edgeAttributes[metricName]).filter(x => !!x)
		const attributeType = attributeTypes.edges[metricName]

		const concentrated = { incoming: [], outgoing: [] }
		metricValues.forEach(element => {
			concentrated.incoming.push(element.incoming)
			concentrated.outgoing.push(element.outgoing)
		})

		if (metricValues.length == 0) {
			return { incoming: 0, outgoing: 0 }
		}

		switch (attributeType) {
			case AttributeTypeValue.relative:
				return { incoming: this.median(concentrated.incoming), outgoing: this.median(concentrated.outgoing) }
			case AttributeTypeValue.absolute:
			default:
				return {
					incoming: concentrated.incoming.reduce((pS, a) => pS + a),
					outgoing: concentrated.outgoing.reduce((pS, a) => pS + a)
				}
		}
	}

	private static median(numbers: number[]): number {
		const middle = (numbers.length - 1) / 2
		numbers.sort((a, b) => a - b)
		return (numbers[Math.floor(middle)] + numbers[Math.ceil(middle)]) / 2
	}
}
