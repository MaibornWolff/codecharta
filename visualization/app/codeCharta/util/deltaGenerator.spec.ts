import { DeltaGenerator } from "./deltaGenerator"
import { TEST_DELTA_MAP_A, TEST_DELTA_MAP_B } from "./dataMocks"
import { CCFile, NodeType } from "../codeCharta.model"
import { NodeDecorator } from "./nodeDecorator"
import { clone } from "./clone"

describe("deltaGenerator", () => {
	let fileA: CCFile
	let fileB: CCFile

	beforeEach(() => {
		fileA = clone(TEST_DELTA_MAP_A)
		fileB = clone(TEST_DELTA_MAP_B)
	})

	it("golden test", () => {
		NodeDecorator.decorateMapWithPathAttribute(fileA)
		NodeDecorator.decorateMapWithPathAttribute(fileB)

		fileA.map.children.push({
			name: "onlyA",
			type: NodeType.FOLDER,
			attributes: {},
			path: "/root/onlyA",
			isExcluded: false,
			isFlattened: false,
			children: [
				{
					name: "special",
					type: NodeType.FOLDER,
					attributes: {},
					path: "/root/onlyA/special",
					isExcluded: false,
					isFlattened: false,
					children: [
						{
							name: "unicorn",
							type: NodeType.FILE,
							attributes: { special: 42 },
							path: "/root/onlyA/special/unicorn",
							isExcluded: false,
							isFlattened: false
						}
					]
				}
			]
		})

		fileB.map.children.push({
			name: "onlyA",
			type: NodeType.FOLDER,
			attributes: {},
			path: "/root/onlyA",
			isExcluded: false,
			isFlattened: false,
			children: [
				{
					name: "special",
					type: NodeType.FOLDER,
					attributes: {},
					path: "/root/onlyA/special",
					isExcluded: false,
					isFlattened: false,
					children: [
						{
							name: "Narwal",
							type: NodeType.FILE,
							attributes: { monster: 666 },
							path: "/root/onlyA/special/Narwal",
							isExcluded: false,
							isFlattened: false
						}
					]
				}
			]
		})

		const result = DeltaGenerator.getDeltaFile(fileA, fileB)

		expect(result.map).toMatchSnapshot()
	})

	it("getDeltaFile should result in expected deltaFiles", () => {
		NodeDecorator.decorateMapWithPathAttribute(fileA)
		NodeDecorator.decorateMapWithPathAttribute(fileB)

		const result = DeltaGenerator.getDeltaFile(fileA, fileB)

		expect(result.map).toMatchSnapshot()
	})

	it("should sum the size of the comparison and reference File", () => {
		NodeDecorator.decorateMapWithPathAttribute(fileA)
		NodeDecorator.decorateMapWithPathAttribute(fileB)

		const result = DeltaGenerator.getDeltaFile(fileA, fileB)

		expect(result.fileMeta.exportedFileSize).toBe(fileA.fileMeta.exportedFileSize + fileB.fileMeta.exportedFileSize)
	})

	it("checking delta calculation between two attribute lists", () => {
		const a = { a: 100, b: 10, c: 1 }
		const b = { a: 110, b: 11, c: 0 }
		const c = { a: 110, b: 11, c: 0, d: 10 }
		const d = { a: 110, b: 11 }
		// eslint-disable-next-line unicorn/prevent-abbreviations
		const e = { d: 110, e: 11 }

		const ab = DeltaGenerator["getDeltaAttributeList"](a, b)
		expect(ab.a).toBe(b.a - a.a)
		expect(ab.b).toBe(b.b - a.b)
		expect(ab.c).toBe(b.c - a.c)

		const ac = DeltaGenerator["getDeltaAttributeList"](a, c)
		expect(ac.a).toBe(c.a - a.a)
		expect(ac.b).toBe(c.b - a.b)
		expect(ac.c).toBe(c.c - a.c)
		expect(ac.d).toBe(c.d)

		const ad = DeltaGenerator["getDeltaAttributeList"](a, d)
		expect(ad.a).toBe(d.a - a.a)
		expect(ad.b).toBe(d.b - a.b)
		expect(ad.c).toBe(-a.c)

		const ae = DeltaGenerator["getDeltaAttributeList"](a, e)
		expect(ae.a).toBe(-a.a)
		expect(ae.b).toBe(-a.b)
		expect(ae.c).toBe(-a.c)
		expect(ae.d).toBe(e.d)
		expect(ae.e).toBe(e.e)
	})
})
