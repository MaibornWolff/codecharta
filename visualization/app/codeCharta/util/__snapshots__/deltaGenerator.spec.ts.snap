// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`deltaGenerator getDeltaFile should result in expected deltaFiles 1`] = `
Object {
  "attributes": Object {},
  "children": Array [
    Object {
      "attributes": Object {
        "functions": 10,
        "mcc": 1,
        "rloc": 20,
      },
      "deltas": Object {
        "functions": 0,
        "mcc": 0,
        "rloc": -80,
      },
      "isExcluded": false,
      "isFlattened": false,
      "link": "http://www.google.de",
      "name": "big leaf",
      "path": "/root/big leaf",
      "type": "File",
    },
    Object {
      "attributes": Object {
        "functions": 11,
        "mcc": 5,
        "rloc": 10,
      },
      "deltas": Object {
        "functions": 11,
        "mcc": 5,
        "rloc": 10,
      },
      "isExcluded": false,
      "isFlattened": false,
      "link": "http://www.google.de",
      "name": "additional leaf",
      "path": "/root/additional leaf",
      "type": "File",
    },
    Object {
      "attributes": Object {},
      "children": Array [
        Object {
          "attributes": Object {
            "functions": 100,
            "mcc": 100,
            "more": 20,
            "rloc": 30,
          },
          "deltas": Object {
            "functions": 0,
            "mcc": 0,
            "more": 20,
            "rloc": 0,
          },
          "isExcluded": false,
          "isFlattened": false,
          "name": "small leaf",
          "path": "/root/Parent Leaf/small leaf",
          "type": "File",
        },
        Object {
          "attributes": Object {
            "functions": 1000,
            "rloc": 70,
          },
          "deltas": Object {
            "functions": 0,
            "mcc": -10,
            "rloc": 0,
          },
          "isExcluded": false,
          "isFlattened": false,
          "name": "other small leaf",
          "path": "/root/Parent Leaf/other small leaf",
          "type": "File",
        },
        Object {
          "attributes": Object {
            "functions": 10,
            "mcc": 1,
            "rloc": 20,
          },
          "deltas": Object {
            "functions": 10,
            "mcc": 1,
            "rloc": 20,
          },
          "isExcluded": false,
          "isFlattened": false,
          "link": "http://www.google.de",
          "name": "big leaf",
          "path": "/root/Parent Leaf/big leaf",
          "type": "File",
        },
      ],
      "deltas": Object {},
      "isExcluded": false,
      "isFlattened": false,
      "name": "Parent Leaf",
      "path": "/root/Parent Leaf",
      "type": "Folder",
    },
  ],
  "deltas": Object {},
  "isExcluded": false,
  "isFlattened": false,
  "name": "root",
  "path": "/root",
  "type": "Folder",
}
`;

exports[`deltaGenerator golden test 1`] = `
Object {
  "attributes": Object {},
  "children": Array [
    Object {
      "attributes": Object {
        "functions": 10,
        "mcc": 1,
        "rloc": 20,
      },
      "deltas": Object {
        "functions": 0,
        "mcc": 0,
        "rloc": -80,
      },
      "isExcluded": false,
      "isFlattened": false,
      "link": "http://www.google.de",
      "name": "big leaf",
      "path": "/root/big leaf",
      "type": "File",
    },
    Object {
      "attributes": Object {
        "functions": 11,
        "mcc": 5,
        "rloc": 10,
      },
      "deltas": Object {
        "functions": 11,
        "mcc": 5,
        "rloc": 10,
      },
      "isExcluded": false,
      "isFlattened": false,
      "link": "http://www.google.de",
      "name": "additional leaf",
      "path": "/root/additional leaf",
      "type": "File",
    },
    Object {
      "attributes": Object {},
      "children": Array [
        Object {
          "attributes": Object {
            "functions": 100,
            "mcc": 100,
            "more": 20,
            "rloc": 30,
          },
          "deltas": Object {
            "functions": 0,
            "mcc": 0,
            "more": 20,
            "rloc": 0,
          },
          "isExcluded": false,
          "isFlattened": false,
          "name": "small leaf",
          "path": "/root/Parent Leaf/small leaf",
          "type": "File",
        },
        Object {
          "attributes": Object {
            "functions": 1000,
            "rloc": 70,
          },
          "deltas": Object {
            "functions": 0,
            "mcc": -10,
            "rloc": 0,
          },
          "isExcluded": false,
          "isFlattened": false,
          "name": "other small leaf",
          "path": "/root/Parent Leaf/other small leaf",
          "type": "File",
        },
        Object {
          "attributes": Object {
            "functions": 10,
            "mcc": 1,
            "rloc": 20,
          },
          "deltas": Object {
            "functions": 10,
            "mcc": 1,
            "rloc": 20,
          },
          "isExcluded": false,
          "isFlattened": false,
          "link": "http://www.google.de",
          "name": "big leaf",
          "path": "/root/Parent Leaf/big leaf",
          "type": "File",
        },
      ],
      "deltas": Object {},
      "isExcluded": false,
      "isFlattened": false,
      "name": "Parent Leaf",
      "path": "/root/Parent Leaf",
      "type": "Folder",
    },
    Object {
      "attributes": Object {},
      "children": Array [
        Object {
          "attributes": Object {},
          "children": Array [
            Object {
              "attributes": Object {
                "monster": 666,
              },
              "deltas": Object {
                "monster": 666,
              },
              "isExcluded": false,
              "isFlattened": false,
              "name": "Narwal",
              "path": "/root/onlyA/special/Narwal",
              "type": "File",
            },
            Object {
              "attributes": Object {
                "special": 42,
              },
              "deltas": Object {
                "special": -42,
              },
              "isExcluded": false,
              "isFlattened": false,
              "name": "unicorn",
              "path": "/root/onlyA/special/unicorn",
              "type": "File",
            },
          ],
          "deltas": Object {},
          "isExcluded": false,
          "isFlattened": false,
          "name": "special",
          "path": "/root/onlyA/special",
          "type": "Folder",
        },
      ],
      "deltas": Object {},
      "isExcluded": false,
      "isFlattened": false,
      "name": "onlyA",
      "path": "/root/onlyA",
      "type": "Folder",
    },
  ],
  "deltas": Object {},
  "isExcluded": false,
  "isFlattened": false,
  "name": "root",
  "path": "/root",
  "type": "Folder",
}
`;
